stl/inc/algorithm

```diff
@@ -2214,6 +2221,285 @@ namespace ranges {
+                return _Last - _First;
+            } else {
+                return _Not_a_difference{};
+            }
+        }
+
+        template <class _Rng>
+        [[nodiscard]] static constexpr auto _Distance_helper(_Rng&& _Range) {
+            _STL_INTERNAL_STATIC_ASSERT(range<_Rng>);
+            if constexpr (sized_range<_Rng>) {
+                return _RANGES distance(_Range);
+            } else {
+                return _Not_a_difference{};
+            }
+        }
+
+        template <class _Ty, class _It1, class _It2, class _Pr, class _Pj1, class _Pj2>
+        [[nodiscard]] static constexpr bool _Match_backwards(
+            const _Ty _First1, _It1 _Last1, const _It2 _First2, _It2 _Last2, _Pr _Pred, _Pj1 _Proj1, _Pj2 _Proj2) {
+            _STL_INTERNAL_STATIC_ASSERT(same_as<_Ty, _It1> || same_as<_Ty, unreachable_sentinel_t>);
+            _STL_INTERNAL_STATIC_ASSERT(bidirectional_iterator<_It1>);
+            _STL_INTERNAL_STATIC_ASSERT(bidirectional_iterator<_It2>);
+            _STL_INTERNAL_STATIC_ASSERT(indirectly_comparable<_It1, _It2, _Pr, _Pj1, _Pj2>);
+
+            for (;;) {
+                if (_First2 == _Last2) { // needle exhausted - match
+                    return true;
+                }
+
+                if (_First1 == _Last1) { // haystack exhausted - no match
+                    return false;
+                }
+
+                --_Last1;
+                --_Last2;
+                if (!_STD invoke(_Pred, _STD invoke(_Proj1, *_Last1), _STD invoke(_Proj2, *_Last2))) {
+                    return false; // non-equal elements - no match
+                }
+            }
+        }
+
+        template <class _It1, class _Se1, class _Diff1, class _It2, class _Pr, class _Pj1, class _Pj2>
+        [[nodiscard]] static constexpr bool _Ends_with_sized_needle(_It1 _First1, _Se1 _Last1, _Diff1 _Count1,
+            _It2 _First2, iter_difference_t<_It2> _Count2, _Pr _Pred, _Pj1 _Proj1, _Pj2 _Proj2) {
+            _STL_INTERNAL_STATIC_ASSERT(sentinel_for<_Se1, _It1>);
+            _STL_INTERNAL_STATIC_ASSERT(_Is_any_of_v<_Diff1, iter_difference_t<_It1>, _Not_a_difference>);
+            _STL_INTERNAL_STATIC_ASSERT(input_iterator<_It2>);
+            _STL_INTERNAL_STATIC_ASSERT(indirectly_comparable<_It1, _It2, _Pr, _Pj1, _Pj2>);
+
+            constexpr bool _Sized1 = !same_as<_Diff1, _Not_a_difference>;
+
+            if constexpr (random_access_iterator<_It1> && _Sized1) {
+                _First1 += static_cast<iter_difference_t<_It1>>(_Count1 - _Count2);
+            } else if constexpr (_Bidi_common<_It1, _Se1>) {
+                if constexpr (_Sized1) {
+                    if ((_Count1 >> 1) >= _Count2) { // beginning of potential match is closer to _Last1
+                        _RANGES advance(_Last1, static_cast<iter_difference_t<_It1>>(-_Count2));
+                        _First1 = _STD move(_Last1);
+                    } else { // beginning of potential match is closer to _First1

```

Comment:
Are all of these cases exercised by the test code?

Reply:
They are now.